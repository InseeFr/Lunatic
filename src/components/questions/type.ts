import { ReactNode } from 'react-markdown/lib/ast-to-react';

export type MessageType = {
	text?: ReactNode;
};
