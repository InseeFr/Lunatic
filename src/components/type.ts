import {
	LunaticComponentDefinition,
	LunaticError,
	LunaticExpression,
	LunaticState,
} from '../use-lunatic/type';
import { CSSProperties, FunctionComponent, ReactNode } from 'react';
import { SuggesterStatus } from '../use-lunatic/use-suggesters';
import useLunatic from '../use-lunatic';

export type VtlExpression = {
	value: string;
	type: 'VTL' | 'VTL|MD';
};

export type LunaticBaseProps<ValueType = unknown> = {
	id: string;
	handleChange: (
		response: { name: string },
		value: ValueType,
		args?: Record<string, unknown>
	) => void;
	errors?: { [id: string]: LunaticError[] };
	preferences?: LunaticState['preferences'];
	declarations?: {
		id: string;
		declarationType: string;
		position: string;
		label: ReactNode;
	}[];
	label?: ReactNode;
	disabled?: boolean;
	missing?: unknown;
	missingResponse?: { name: string; value?: unknown };
	management?: LunaticState['management'];
	description?: ReactNode;
	shortcut?: boolean;
	required?: boolean;
	value: null | ValueType;
	readOnly?: boolean;
	className?: string;
	style?: CSSProperties;
	iteration?: number;
	executeExpression: LunaticState['executeExpression'];
	features?: string[];
	componentType?: string;
};

export type SuggesterOption = {
	children?: string[];
	id?: string;
	description?: ReactNode;
	label?: ReactNode;
	value: string;
	niveau?: string;
	parent?: string;
	tokensMap?: Record<string, { count: number; fields: string[] }>;
};

type ComponentPropsByType = {
	InputNumber: LunaticBaseProps<number | null> & {
		min: number;
		max: number;
		decimals: number;
		unit?: string;
		response: { name: string };
	};
	Input: LunaticBaseProps<string> & {
		maxLength?: number;
		value: null | string;
		response: { name: string };
	};
	Sequence: Pick<
		LunaticBaseProps<string>,
		'id' | 'declarations' | 'label' | 'style'
	>;
	Subsequence: Pick<LunaticBaseProps<string>, 'id' | 'declarations' | 'label'>;
	Question: Pick<LunaticBaseProps<unknown>, 'label' | 'description'>;
	ComponentSet: LunaticBaseProps<unknown> & {
		components: LunaticComponentDefinition[];
		value: Record<string, unknown>;
		response: undefined;
	};
	RosterForLoop: LunaticBaseProps<unknown> & {
		lines: { min: number; max: number };
		iterations?: number;
		components: LunaticComponentDefinition[];
		executeExpression: LunaticState['executeExpression'];
		value: Record<string, unknown[]>;
		headers?: Array<{ label: ReactNode }>;
		paginatedLoop?: boolean;
	};
	Loop: LunaticBaseProps<unknown> & {
		lines: { min: number; max: number };
		iterations?: number;
		components: LunaticComponentDefinition[];
		executeExpression: LunaticState['executeExpression'];
		value: Record<string, unknown[]>;
		headers?: Array<{ label: ReactNode }>;
		paginatedLoop?: boolean;
	};
	Table: LunaticBaseProps<unknown> & {
		value: Record<string, unknown>;
		header: Array<{
			label: ReactNode;
			rowspan?: number;
			colspan?: number;
		}>;
		body: Array<Array<{ label: LunaticExpression }>>;
		executeExpression: LunaticState['executeExpression'];
		iteration: LunaticState['pager']['iteration'];
	};
	Datepicker: LunaticBaseProps<string | null> & {
		min?: string;
		max?: string;
		response: { name: string };
	};
	CheckboxGroup: LunaticBaseProps<Record<string, boolean | null>> & {
		responses: Array<{
			id: string;
			label: ReactNode;
			description?: ReactNode;
			response: { name: string };
		}>;
		handleChange: (
			response: { name: string },
			value: boolean,
			args?: Record<string, unknown>
		) => void;
	};
	CheckboxOne: LunaticBaseProps<string | null> & {
		options: Array<{
			description: ReactNode;
			label: ReactNode;
			value: string;
		}>;
		response: { name: string };
	};
	Switch: LunaticBaseProps<boolean> & {
		response: { name: string };
		statusLabel?: { true: string; false: string };
	};
	CheckboxBoolean: LunaticBaseProps<boolean> & {
		options: Array<{ description: ReactNode; label: ReactNode; value: string }>;
		response: { name: string };
	};
	Radio: LunaticBaseProps<string | null> & {
		options: Array<{ description: ReactNode; label: ReactNode; value: string }>;
		checkboxStyle?: boolean;
		response: { name: string };
	};
	Roundabout: LunaticBaseProps<string> & {
		iterations: number;
		goToPage: ReturnType<typeof useLunatic>['goToPage'];
		page: string;
		locked?: boolean;
		expressions: {
			unnecessary?: Array<boolean>;
			complete?: Array<boolean>;
			partial?: Array<boolean>;
			label?: Array<string>;
		};
	};
	Dropdown: LunaticBaseProps<string | null> & {
		options: Array<{ description: ReactNode; label: ReactNode; value: string }>;
		response: { name: string };
		writable?: boolean;
	};
	Textarea: LunaticBaseProps<string> & {
		cols?: number;
		placeHolder?: string;
		maxLength?: number;
		rows?: number;
		response: { name: string };
	};
	FilterDescription: Pick<LunaticBaseProps<string>, 'id' | 'label'>;
	Accordion: Pick<LunaticBaseProps<string>, 'id' | 'label' | 'description'> & { bgColor?: string };
	PairwiseLinks: Omit<LunaticBaseProps, 'value'> & {
		components: LunaticComponentDefinition[];
		features?: LunaticState['features'];
		executeExpression: LunaticState['executeExpression'];
		xAxisIterations: number;
		yAxisIterations: number;
		symLinks: Record<string, Record<string, string>>;
		value: Record<string, unknown[]>;
	};
	Suggester: LunaticBaseProps<string | null> & {
		storeName: string;
		getSuggesterStatus: (name: string) => {
			status: SuggesterStatus;
			timestamp: number;
		};
		optionRenderer: FunctionComponent<{
			option: SuggesterOption;
			placeholder?: string;
			search?: string;
		}>;
		labelRenderer: FunctionComponent<{
			option?: SuggesterOption;
			selected?: boolean;
			search?: string;
		}>;
		idbVersion?: string;
		focused: boolean;
		response: { name: string };
	};
	Summary: LunaticBaseProps<string | null> & {
		sections: Array<{
			responses?: Array<{ label: VtlExpression; value: VtlExpression }>;
			title?: VtlExpression;
			iterations?: number;
		}>
	};
};

export type LunaticComponentType = keyof ComponentPropsByType;

export type LunaticComponentProps<
	T extends LunaticComponentType = LunaticComponentType
> = ComponentPropsByType[T];
