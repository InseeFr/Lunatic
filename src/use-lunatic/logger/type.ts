type LoggerMessage =
	| {
			type: 'ERROR';
			error: Error;
	  }
	| {
			type: 'WARNING';
			message: string;
	  };

export type LunaticLogger = (msg: LoggerMessage) => void;
