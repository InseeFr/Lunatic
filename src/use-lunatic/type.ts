import type { ReactNode } from 'react';
import type {
	ComponentType,
	ControlType,
	LunaticSource,
	Variable,
} from './type-source';
import { SuggesterStatus } from './use-suggesters';
import type { LunaticVariablesStore } from './commons/variables/lunatic-variables-store';

export type LunaticComponentDefinition<
	T extends ComponentType['componentType'] = ComponentType['componentType']
> = ComponentType & { componentType: T };
export type LunaticControl = ControlType;

export type VTLBindings = { [variableName: string]: unknown };

export type LunaticData = Partial<
	Record<Exclude<VariableType, 'COLLECTED'>, Record<string, unknown>> & {
		COLLECTED: Record<string, LunaticCollectedValue>;
	}
>;

export type LunaticValues = {
	[variableName: string]: unknown;
};

export type LunaticError = Pick<
	ControlType,
	'id' | 'criticality' | 'typeOfControl'
> & {
	errorMessage: ReactNode;
};

export type VariableType = 'COLLECTED' | 'EXTERNAL' | 'CALCULATED';
export type ExpressionType = 'VTL' | 'VTL|MD';
export type LunaticExpression = {
	value: string;
	type: ExpressionType;
	bindingDependencies?: string[];
};
export type TODO = unknown; // Temporary type to mark types as unresolved

export type LunaticVariable = Variable;
export type LunaticCollectedValue = {
	COLLECTED: unknown;
	EDITED: unknown;
	FORCED: unknown;
	INPUTED: unknown;
	PREVIOUS: unknown;
};

export type LunaticOverviewItem = {
	id: string;
	page: string;
	type: string;
	evaluatedLabel: string;
	visible: boolean;
	reached: boolean;
	parent?: unknown;
	label: LunaticExpression;
	conditionFilter?: {
		bindingDependencies?: string[];
	};
	children: LunaticOverviewItem[];
};

// We need a mapped type to correlate type and variableType
export type LunaticStateVariable = {
	[key in LunaticVariable['variableType']]: {
		type: key;
		value: unknown;
		variable: LunaticVariable & { variableType: key };
		CalculatedLinked?: LunaticVariable[];
	};
}[LunaticVariable['variableType']];

export type LunaticState = {
	updatedAt: number;
	variables: LunaticVariablesStore;
	pages: {
		[key: number | string]:
			| {
					components: ComponentType[];
					isLoop: false;
					iterations?: undefined;
					loopDependencies?: undefined;
					subPages?: undefined;
			  }
			| {
					components: ComponentType[];
					isLoop: true;
					iterations: ExpressionType;
					// Variables affecting this loop
					loopDependencies: string[];
					// List of child pages (ex: ['20.1', '20.2']
					subPages: string[];
			  };
	};
	isInLoop: boolean;
	isFirstPage: boolean;
	isLastPage: boolean;
	features: ['VTL'] | ['VTL', 'MD'];
	preferences: ['COLLECTED'];
	savingType: 'COLLECTED';
	// Map of variable associated with the expression used to repopulate it
	cleaning: LunaticSource['cleaning'];
	// Map of variable with the missing variable names associated with it ex: {ADR_COLL: ['ADR_COLL_MISSING']}
	missingBlock: { [variable: string]: string[] };
	// Map of variable with the missing variable names associated with it ex: {ADR_COLL: ['ADR_COLL_MISSING']}
	resizing: {
		[variable: string]: {
			// VTL expression
			size: string;
			// List of variables that will need resizing
			variables: string[];
			sizeForLinksVariables?: unknown;
			linksVariables?: string[];
		};
	};
	overview: LunaticOverviewItem[];
	pager: {
		lastReachedPage?: string;
		maxPage: string;
		nbSubPages?: number;
		page: string;
		subPage?: number;
		// Iteration index (starting at 0)
		iteration?: number;
		nbIterations?: number;
		shallowIteration?: number;
		linksIterations?: number[];
	};
	// TODO : Explain this
	waiting: boolean;
	// Errors for the form
	errors?: { [page: string]: { [id: string]: LunaticError[] } };
	// Contains the errors for the current page / iteration
	currentErrors?: { [id: string]: LunaticError[] };
	// Errors
	modalErrors?: Record<string, LunaticError[]>;
	// Handler to call when updating a value
	handleChange: (
		response: { name: string },
		value: any,
		args?: Record<string, unknown>
	) => void;
	// Run and expression using the value from the state
	executeExpression: <T extends unknown = unknown>(
		expression: unknown,
		args?: {
			iteration?: number;
			// @deprecated
			bindingDependencies?: string[];
			deps?: string[];
		}
	) => T;
	// Update the value collected for the variable
	updateBindings: (variableName: string, value: unknown) => unknown;
	// Enable controls for data (form validation)
	activeControls: boolean;
	// enable shortcut on radio/checkbox/missing buttons
	shortcut?: boolean;
	// TODO : Explain this
	management?: boolean;
	goToPage: (page: {
		page: string;
		iteration?: number;
		nbIterations?: number;
		subPage?: number;
	}) => void;
	// Enable components to independently navigate next/previous
	goNextPage: () => void;
	goPreviousPage: () => void;
	getSuggesterStatus: (name: string) => {
		status: SuggesterStatus;
		timestamp: number;
	};
};
