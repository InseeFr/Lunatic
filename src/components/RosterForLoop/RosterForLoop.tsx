import { Fragment, useCallback, useState } from 'react';
import type { LunaticComponentProps } from '../type';
import { Table, Tbody, Td, Tr, TableHeader } from '../shared/Table';
import { times } from '../../utils/array';
import { LunaticComponents } from '../LunaticComponents';
import { blockedInLoopComponents } from '../Loop/constant';
import {
	ComponentErrors,
	getComponentErrors,
} from '../shared/ComponentErrors/ComponentErrors';
import { CustomLoop } from '../Loop/Loop';
import { objectMap } from '../../utils/object';

const DEFAULT_MIN_ROWS = 1;
const DEFAULT_MAX_ROWS = 12;

/**
 * Loop displayed as a table
 */
export const RosterForLoop = (
	props: LunaticComponentProps<'RosterForLoop'>
) => {
	const {
		value: valueMap,
		lines,
		errors,
		handleChanges,
		declarations,
		header,
		iterations,
		id,
		getComponents,
		label,
		...otherProps // These props will be passed down to the child components
	} = props;
	const min = lines?.min ?? DEFAULT_MIN_ROWS;
	const max = lines?.max ?? DEFAULT_MAX_ROWS;
	const [nbRows, setNbRows] = useState(Math.max(min, iterations));

	const addRow = useCallback(() => {
		if (nbRows < max) {
			setNbRows(nbRows + 1);
		}
	}, [max, nbRows]);

	const removeRow = useCallback(() => {
		if (nbRows <= min) {
			return;
		}
		const newNbRows = nbRows - 1;
		setNbRows(newNbRows);
		// Downsize all variables by 1
		const newResponses = Object.entries(valueMap).map(([k, v]) => {
			return {
				name: k,
				value: v?.filter((_, i) => i < newNbRows),
			};
		});
		handleChanges(newResponses);
	}, [nbRows, min, valueMap, handleChanges]);

	if (nbRows === 0) {
		return null;
	}

	let cols = 0;

	return (
		<CustomLoop
			{...props}
			errors={getComponentErrors(errors, props.id)}
			addRow={nbRows === max ? undefined : addRow}
			removeRow={nbRows === min ? undefined : removeRow}
			canControlRows={!!(min && max && min !== max)}
		>
			<Table id={id}>
				{header && <TableHeader header={header} />}
				<Tbody>
					{times(nbRows, (n) => {
						const components = getComponents(n);
						const lineErrors = getComponentErrors(props.errors, `${id}-${n}`);
						const hasLineErrors = !!lineErrors?.length;
						cols = components.length;
						return (
							<Fragment key={n}>
								<Tr
									row={n}
									key={n}
									className={
										hasLineErrors ? 'lunatic-row-has-error' : undefined
									}
								>
									<LunaticComponents
										blocklist={blockedInLoopComponents}
										components={components}
										memo
										componentProps={(c) => ({
											...otherProps,
											...c,
											id: `${c.id}-${n}`,
											errors,
										})}
										wrapper={({ children }) => <Td>{children}</Td>}
									/>
								</Tr>
								{hasLineErrors && (
									<Tr className="lunatic-errors">
										<Td colSpan={cols}>
											<ComponentErrors errors={lineErrors} />
										</Td>
									</Tr>
								)}
							</Fragment>
						);
					})}
				</Tbody>
			</Table>
		</CustomLoop>
	);
};
