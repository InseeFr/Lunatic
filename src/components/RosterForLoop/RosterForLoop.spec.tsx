import { beforeEach, describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { RosterForLoop } from './RosterForLoop';
import type { LunaticComponentProps } from '../type';

describe('RosterForLoop', () => {
	const mockOnChange = vi.fn();

	beforeEach(() => {
		mockOnChange.mockClear();
	});

	const getComponents = (iteration: number) => [
		{
			componentType: 'Input',
			maxLength: 249,
			id: 'jrc3ye5q-QOP-lo6tcvvx',
			response: {
				name: 'TABDYN1',
			},
			iteration: iteration,
			value: 'azeaze',
		} as LunaticComponentProps<'Input'>,
	];

	it('renders the right number of columns', () => {
		const { container } = render(
			<RosterForLoop
				value={{ name: ['John Doe', 'Jane Doe'] }}
				handleChanges={mockOnChange}
				label="Ceci est un test"
				id="table"
				lines={{ min: 1, max: 3 }}
				iterations={2}
				getComponents={getComponents}
				executeExpression={() => null as any}
			/>
		);
		expect(container).toMatchSnapshot();
		expect(screen.getAllByRole('row')).toHaveLength(2);
	});
});
