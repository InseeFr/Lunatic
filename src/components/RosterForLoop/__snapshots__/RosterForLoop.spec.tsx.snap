// Vitest Snapshot v1

exports[`RosterForLoop > renders the right number of columns 1`] = `
<div>
  <label
    class="lunatic-label"
    for="table"
    id="label-table"
  >
    Ceci est un test
  </label>
  <table
    class="lunatic-table"
    id="table-table"
  >
    <tbody
      class="lunatic-table-tbody"
    >
      <tr
        class="lunatic-table-tr"
      >
        <td
          class="lunatic-table-td"
        >
          <div
            class="field-container"
          >
            <div
              class="field"
            >
              <div
                class="lunatic-input"
              >
                <input
                  aria-invalid="false"
                  aria-labelledby="label-jrc3ye5q-QOP-lo6tcvvx-0"
                  autocomplete="off"
                  id="jrc3ye5q-QOP-lo6tcvvx-0"
                  maxlength="249"
                  type="text"
                  value="azeaze"
                />
              </div>
            </div>
          </div>
        </td>
      </tr>
      <tr
        class="lunatic-table-tr"
      >
        <td
          class="lunatic-table-td"
        >
          <div
            class="field-container"
          >
            <div
              class="field"
            >
              <div
                class="lunatic-input"
              >
                <input
                  aria-invalid="false"
                  aria-labelledby="label-jrc3ye5q-QOP-lo6tcvvx-1"
                  autocomplete="off"
                  id="jrc3ye5q-QOP-lo6tcvvx-1"
                  maxlength="249"
                  type="text"
                  value="azeaze"
                />
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <input
    class="button-lunatic"
    type="button"
    value="Add row"
  />
  <input
    class="button-lunatic"
    type="button"
    value="Remove row"
  />
</div>
`;
